neqo-crypto/src/agent.rs

```diff
@@ -51,6 +51,146 @@ use crate::{
     time::{Time, TimeHolder},
 };
 
+/// Private trait for Certificate Compression implementation
+/// # Safety
+///
+/// Use `SafeCertCompression` to implement an encoder/decoder instead.
+unsafe trait UnsafeCertCompression {
+    unsafe extern "C" fn decode_callback(
+        input: *const ssl::SECItem,
+        output: *mut ::std::os::raw::c_uchar,
+        output_len: usize,
+        used_len: *mut usize,
+    ) -> ssl::SECStatus;
+
+    unsafe extern "C" fn encode_callback(
+        input: *const ssl::SECItem,
+        output: *mut ssl::SECItem,
+    ) -> ssl::SECStatus;
+}
+
+/// The trait is used to represent a certificate compression data structure
+/// Used in order to enable Certificate Compression extension during TLS connection
+pub trait CertificateCompressor {
+    /// Certificate Compression identifier as in RFC8879
+    const ID: u16;
+    /// Certification Compression name (used only for logging/debugging)
+    const NAME: &CStr;
+    /// Certificate Compression could be used to encode and decode a certificate
+    /// though the encoding is not frequently used
+    /// Enable decoding field is used to signal to the implementation
+    /// to use the encoding as well
+    const ENABLE_ENCODING: bool = false;
+
+    /// Certificate Compression encoding function
+    /// If the implementation is not provided, we only copy the data
+    /// NB: If `ENABLE_ENCODING` is not set, the function pointer provided to NSS will be null
+    /// # Errors
+    /// Encoding was unsuccessful, for example, not enough memory
+    fn encode(input: &[u8], output: &mut [u8]) -> Res<usize> {
+        let len = std::cmp::min(input.len(), output.len());
+        output[..len].copy_from_slice(&input[..len]);
+        Ok(len)
+    }
+
+    /// Certificate Compression decoding function. Returns the length of the decoded buffer,
+    /// or 0 if an error has occured.
+    /// # Errors
+    /// Decoding was unsuccessful, for example, not enough memory
+    fn decode(input: &[u8], output: &mut [u8]) -> Res<usize>;
+}
+
+/// The trait is responsible for calling `CertificateCompression` encoding and decoding
+/// functions using the NSS types
+unsafe impl<T: CertificateCompressor> UnsafeCertCompression for T {
+    unsafe extern "C" fn decode_callback(
+        input: *const ssl::SECItem,
+        output: *mut ::std::os::raw::c_uchar,
+        output_len: usize,
+        used_len: *mut usize,
+    ) -> ssl::SECStatus {
+        unsafe {
```

Comment:
I prefer to focus the unsafe blocks on the code that is genuinely unsafe.  In this case, that's a much smaller scope than this (lines 116, 120, and 121 only in my current diff view).  Can you move the block to that smaller scope?